go/arrow/flight/flightsql/driver.go

```diff
@@ -0,0 +1,779 @@
+	}
+
+	if d.txn != nil && d.txn.txn.IsValid() {
+		return ErrTransactionInProgress
+	}
+
+	// Drivers must ensure all network calls made by Close
+	// do not block indefinitely (e.g. apply a timeout).
+	err := d.client.Close()
+	d.client = nil
+
+	return err
+}
+
+// Begin starts and returns a new transaction.
+func (d *Driver) Begin() (driver.Tx, error) {
+	return d.BeginTx(context.Background(), sql.TxOptions{})
+}
+
+func (d *Driver) BeginTx(ctx context.Context, opts sql.TxOptions) (driver.Tx, error) {
+	tx, err := d.client.BeginTransaction(ctx)
+	if err != nil {
+		return nil, err
+	}
+	d.txn = tx
+
+	return &Tx{tx: tx, timeout: d.timeout}, nil
+}
+
+func fromArrowType(arr arrow.Array, idx int) (interface{}, error) {
+	switch c := arr.(type) {
+	case *array.Boolean:
+		return c.Value(idx), nil
+	case *array.Float16:
+		return float64(c.Value(idx).Float32()), nil
+	case *array.Float32:
+		return float64(c.Value(idx)), nil
+	case *array.Float64:
+		return c.Value(idx), nil
+	case *array.Int8:
+		return int64(c.Value(idx)), nil
+	case *array.Int16:
+		return int64(c.Value(idx)), nil
+	case *array.Int32:
+		return int64(c.Value(idx)), nil
+	case *array.Int64:
+		return c.Value(idx), nil
+	case *array.String:
+		return c.Value(idx), nil
+	case *array.Time32:
+		dt, ok := arr.DataType().(*arrow.Time32Type)
+		if !ok {
+			return nil, fmt.Errorf("datatype %T not matching time32", arr.DataType())
+		}
+		v := c.Value(idx)
+		return v.ToTime(dt.TimeUnit()), nil
+	case *array.Time64:
+		dt, ok := arr.DataType().(*arrow.Time64Type)
+		if !ok {
+			return nil, fmt.Errorf("datatype %T not matching time64", arr.DataType())
+		}
+		v := c.Value(idx)
+		return v.ToTime(dt.TimeUnit()), nil
+	case *array.Timestamp:
+		dt, ok := arr.DataType().(*arrow.TimestampType)
+		if !ok {
+			return nil, fmt.Errorf("datatype %T not matching timestamp", arr.DataType())
+		}
+		v := c.Value(idx)
+		return v.ToTime(dt.TimeUnit()), nil
+	}
+
+	return nil, fmt.Errorf("type %T: %w", arr, ErrNotSupported)
+}
+
+func toArrowDataType(value interface{}) (arrow.DataType, error) {
+	switch value.(type) {
+	case bool:
+		return &arrow.BooleanType{}, nil
+	case float32:
+		return &arrow.Float32Type{}, nil
+	case float64:
+		return &arrow.Float64Type{}, nil
+	case int8:
+		return &arrow.Int8Type{}, nil
+	case int16:
+		return &arrow.Int16Type{}, nil
+	case int32:
+		return &arrow.Int32Type{}, nil
+	case int64:
+		return &arrow.Int64Type{}, nil
+	case uint8:
+		return &arrow.Uint8Type{}, nil
+	case uint16:
+		return &arrow.Uint16Type{}, nil
+	case uint32:
+		return &arrow.Uint32Type{}, nil
+	case uint64:
+		return &arrow.Uint64Type{}, nil
+	case string:
+		return &arrow.StringType{}, nil
+	case time.Time:
+		return &arrow.Time64Type{Unit: arrow.Nanosecond}, nil
+	}
+	return nil, fmt.Errorf("type %T: %w", value, ErrNotSupported)
+}
+
+func setFieldValue(builder array.Builder, arg interface{}) error {
+	switch b := builder.(type) {
+	case *array.BooleanBuilder:
+		switch v := arg.(type) {
```

Comment:
you also need to handle a union builder

Reply:
Is this a requirement from the start? Do I need to support _all_ builders? Why doesn't arrow provide a helper for doing this? ;-)

Reply:
if you're not going to support *all* value types, then you should at least add a Godoc comment stating that only the types implemented are supported for parameter values.

> Why doesn't arrow provide a helper for doing this? ;-)

It's honestly not a common use case to have an `interface{}` and needing to append it to a `Builder` that isn't already typed. :smile: In the cases where it happens you typically already *know* that the types match so if you type switch on the builder, you can assume the value type or vice versa. The closest case to this would be the way that the `CSV` reader works by having parsing function pointers that it can call which will properly type assert.

As long as what is supported and what is not is explicitly documented for consumers, I'm okay with that.

Reply:
Ok, will add a comment and only support scalar values as this is most probably the only thing needed for parameters. If it turns out we require more, we can still add it later...